File: Facultate/Anul 2/Sem 2/MPP/Laburi/Edi/CSharpServer/Server/services/MainService.cs
using NLog;
using Server.repos;
using Shared.dtos;
using Shared.exceptions;
using Shared.models;
using Shared.services;
using Shared.utils;

namespace Server.services;

public class MainService : IMainService
{
    private readonly IEmployeeRepo _employeeRepo;
    private readonly IRidesRepo _ridesRepo;
    private readonly IReservationsRepo _reservationsRepo;
    private static readonly Logger Logger = LogManager.GetCurrentClassLogger();

    public MainService(IEmployeeRepo employeeRepo, IRidesRepo ridesRepo, IReservationsRepo reservationsRepo)
    {
        _employeeRepo = employeeRepo;
        _ridesRepo = ridesRepo;
        _reservationsRepo = reservationsRepo;
    }

    public List<RideDTO> FindRides(DateOnly rideDate, string departureHour, string destination)
    {
        var departureTime = TimeOnly.Parse(departureHour);
        var foundRide = _ridesRepo.FindAll()
            .FirstOrDefault(r =>
                r.Date == rideDate &&
                r.DepartureTime == departureTime &&
                r.Destination == destination)
            ?? throw new ServiceException("Ride not found");

        var reservations = _reservationsRepo.FindReservationsByRide(foundRide.Id);        
        var result = new List<RideDTO>();
        int seatNo = 1;

        foreach (var reservation in reservations)
        {
            for (int i = 0; i < reservation.ReservedSeatsCount; i++)
                result.Add(new RideDTO(seatNo++, reservation.ClientName));
        }

        while (seatNo <= foundRide.AvailableSeats)
            result.Add(new RideDTO(seatNo++, "-"));

        return result;
    }

    public List<Ride> GetAllRides() => _ridesRepo.FindAll().ToList();

    public void MakeReservation(long rideId, string clientName, int noSeats)
    {
        var reservation = new Reservation
        {
            ClientName = clientName,
            ReservedSeatsCount = noSeats,
            RideId = rideId
        };
        _reservationsRepo.Save(reservation);

        var ride = _ridesRepo.FindOne(rideId);
        if (ride != null)
        {
            if (ride.AvailableSeats - noSeats > 0)
            {
                ride.AvailableSeats -= noSeats;
                _ridesRepo.Update(ride);
            }
            else
            {
                throw new ServiceException("Reservation not found");
            }
        }
        
        Logger.Info($"Reservation made: {clientName}, {noSeats} seats, ride {rideId}");
    }

    public void CancelReservation(long reservationId)
    {
        var reservation = _reservationsRepo.FindOne(reservationId);
        if (reservation == null) return;

        var ride = _ridesRepo.FindOne(reservation.RideId);
        if (ride != null)
        {
            ride.AvailableSeats += reservation.ReservedSeatsCount;
            _ridesRepo.Update(ride);
        }

        _reservationsRepo.Delete(reservationId);
      
        Logger.Info($"Reservation {reservationId} cancelled");
    }

    public Ride? GetRide(long rideId) => _ridesRepo.FindOne(rideId);

    public List<Reservation> FindReservations() => _reservationsRepo.FindAll().ToList();

    public void AddObserver(IObserver observer)
    {
        
    }

    public void RemoveObserver(IObserver observer)
    {
        
    }
}

File: Facultate/Anul 2/Sem 2/MPP/Laburi/Edi/JavaClient C# Server/Server/src/main/java/services/MainService.java
package services;

import dtos.RideDTO;
import models.Reservation;
import models.Ride;
import repos.EmployeeRepo;
import repos.ReservationsRepo;
import repos.RidesRepo;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

import exceptions.ServiceException;
import utils.Observer;

public class MainService implements IMainService {

    private final EmployeeRepo employeeDBRepo;
    private final RidesRepo ridesDBRepo;
    private final ReservationsRepo reservationsDBRepo;
    private final List<Observer>  observers = new CopyOnWriteArrayList<>();

    public MainService(EmployeeRepo employeeDBRepo, RidesRepo ridesDBRepo, ReservationsRepo reservationsDBRepo) {
        this.employeeDBRepo = employeeDBRepo;
        this.ridesDBRepo = ridesDBRepo;
        this.reservationsDBRepo = reservationsDBRepo;
    }


    public List<RideDTO> findRides(LocalDate rideDate, String departureHour, String destination) {
        List<Ride> rides = ridesDBRepo.findAll();
        Ride foundRide = rides.stream()
                .filter(ride -> ride.getDate().equals(rideDate) &&
                        ride.getDepartureTime().equals(LocalTime.parse(departureHour, DateTimeFormatter.ofPattern("HH:mm"))) &&
                        ride.getDestination().equals(destination))

                .findFirst()
                .orElse(null);
        if (foundRide == null) {
            throw new ServiceException("Ride not found");
        }
        List<Reservation> reservations = reservationsDBRepo.findReservationsByRideId(foundRide.getId());
        List<RideDTO> rideDTOList = new ArrayList<>();
        int seatNo = 1;
        for (Reservation reservation : reservations) {
            Integer seatsCount = reservation.getReservedSeatsCount();
            while (seatsCount > 0) {
                rideDTOList.add(new RideDTO(seatNo, reservation.getClientName()));
                seatNo++;
                seatsCount--;
            }
        }
        while (seatNo <= foundRide.getAvailableSeats()) {
            rideDTOList.add(new RideDTO(seatNo, "-"));
            seatNo++;
        }
        return rideDTOList;
    }

    public List<Ride> getAllRides() {
        return ridesDBRepo.findAll();
    }

    public void makeReservation(Long rideId, String text, int noSeats) {
        Reservation reservation = new Reservation(text, noSeats, rideId);
        reservationsDBRepo.save(reservation);
        Ride ride = ridesDBRepo.findOne(rideId).orElse(null);
        if(ride != null) {
            ride.setAvailableSeats(ride.getAvailableSeats() - noSeats);
            ridesDBRepo.update(ride);
        }
        notifyObservers();
    }

    public void cancelReservation(Long reservationId) {
        Reservation reservation = reservationsDBRepo.findOne(reservationId).orElse(null);
        if (reservation != null) {
            ridesDBRepo.findOne(reservation.getRideId()).ifPresent(ride -> {
                ride.setAvailableSeats(ride.getAvailableSeats() + reservation.getReservedSeatsCount());
                ridesDBRepo.update(ride);
            });
            reservationsDBRepo.delete(reservationId);
            notifyObservers();
        }
    }

    public Ride getRide(Long rideId) {
        return ridesDBRepo.findOne(rideId).orElse(null);
    }

    public List<Reservation> findReservations() {
        return reservationsDBRepo.findAll();
    }

    private void notifyObservers() {
        for (Observer observer : observers) {
            new Thread(observer::update).start();
        }
    }

    public void addObserver(Observer observer) {
        observers.add(observer);
    }

    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }
}
